cmd/outputs_cloud.go

```diff
@@ -0,0 +1,178 @@
+	conf, warn, err := cloudapi.GetConsolidatedConfig(
+		test.derivedConfig.Collectors[builtinOutputCloud.String()],
+		gs.Env,
+		"", // Historically used for -o cloud=..., no longer used (deprecated).
+		test.derivedConfig.Options.Cloud,
+		test.derivedConfig.Options.External,
+	)
+	if err != nil {
+		return err
+	}
+
+	if warn != "" {
+		gs.Logger.Warn(warn)
+	}
+
+	// If not, we continue with some validations and the creation of the test run.
+	if err := validateRequiredSystemTags(test.derivedConfig.Options.SystemTags); err != nil {
+		return err
+	}
+
+	if !conf.Name.Valid || conf.Name.String == "" {
+		scriptPath := test.source.URL.String()
+		if scriptPath == "" {
+			// Script from stdin without a name, likely from stdin
+			return errors.New("script name not set, please specify K6_CLOUD_NAME or options.cloud.name")
+		}
+
+		conf.Name = null.StringFrom(filepath.Base(scriptPath))
+	}
+	if conf.Name.String == "-" {
+		conf.Name = null.StringFrom(defaultTestName)
+	}
+
+	thresholds := make(map[string][]string)
+	for name, t := range test.derivedConfig.Thresholds {
+		for _, threshold := range t.Thresholds {
+			thresholds[name] = append(thresholds[name], threshold.Source)
+		}
+	}
+
+	et, err := lib.NewExecutionTuple(
+		test.derivedConfig.Options.ExecutionSegment,
+		test.derivedConfig.Options.ExecutionSegmentSequence,
+	)
+	if err != nil {
+		return err
+	}
+	executionPlan := test.derivedConfig.Options.Scenarios.GetFullExecutionRequirements(et)
+
+	duration, testEnds := lib.GetEndOffset(executionPlan)
+	if !testEnds {
+		return errors.New("tests with unspecified duration are not allowed when outputting data to k6 cloud")
+	}
+
+	if conf.MetricPushConcurrency.Int64 < 1 {
+		return fmt.Errorf("metrics push concurrency must be a positive number but is %d",
+			conf.MetricPushConcurrency.Int64)
+	}
+
+	if conf.MaxTimeSeriesInBatch.Int64 < 1 {
+		return fmt.Errorf("max allowed number of time series in a single batch must be a positive number but is %d",
+			conf.MaxTimeSeriesInBatch.Int64)
+	}
+
+	var testArchive *lib.Archive
+	if !test.derivedConfig.NoArchiveUpload.Bool {
+		testArchive = test.initRunner.MakeArchive()
+	}
+
+	testRun := &cloudapi.TestRun{
+		Name:       conf.Name.String,
+		ProjectID:  conf.ProjectID.Int64,
+		VUsMax:     int64(lib.GetMaxPossibleVUs(executionPlan)), //nolint:gosec
+		Thresholds: thresholds,
+		Duration:   int64(duration / time.Second),
+		Archive:    testArchive,
+	}
+
+	logger := gs.Logger.WithFields(logrus.Fields{"output": builtinOutputCloud.String()})
+
+	apiClient := cloudapi.NewClient(
+		logger, conf.Token.String, conf.Host.String, consts.Version, conf.Timeout.TimeDuration())
+
+	response, err := apiClient.CreateTestRun(testRun)
+	if err != nil {
+		return err
+	}
+
+	// We store the test run id in the environment, so it can be used later.
+	test.preInitState.RuntimeOptions.Env[testRunIDKey] = response.ReferenceID
```

Comment:
@mstoykov While checking if this would be enough, I realized that both:

- `output.Params.Environment`
- and `lib.TestPreInitState.LookupEnv`

rely on `gs.Env`. 

So, If you think it's better, for safety, just in case anybody tries to read that value from any of these, I'll be happy to also populate that environment variable onto `gs.Env`. 

But, for now, I decided to store it only where needed, so we don't start populating everything just because we have different mechanisms to query the environment for legacy reasons, and thus making a potential refactor/deprecation harder to reason about.

Reply:
Created https://github.com/grafana/k6/issues/4145